### DISCUSSION TITLE
How to have Forge find Models in folder/directory.

### DISCUSSION THREAD

**Discussion by LindezaBlue** [lllyasviel/stable-diffusion-webui-forge#320] [Show and tell]
I was having a hard time trying to figure out what to put in the webui-user.bat file since the examples in the folder didn't say you needed quotes for the directory, and didn't say to put the folders right after the first commandline_args. 

So to help I am sharing what I did that worked. just replace the folders shown below with the location yours uses, and make sure the slashes are facing forwards / otherwise it won't work. 

Hope this helps!~

@echo off

set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS= --ckpt-dir "D:/ai/a1111/models/Stable-diffusion" --embeddings-dir "D:/ai/a1111/embeddings" --lora-dir "D:/ai/a1111/models/Lora"

call webui.bat

**Comment by cooperdk** [comment]
Seems you have little knowledge of Python, but that is okay.

The forward slashes could/should be replaced with double backward slashes. That is how you specify directory separators on Windows. So:

set COMMANDLINE_ARGS= --ckpt-dir "D:\\ai\\a1111\\models\\Stable-diffusion" .....

**Reply by LindezaBlue** [reply]
No it's not really a misunderstanding on how python works, for some reason the directories need to have forward slashes. I've tried back slashes as that is how nearly every system does things, but it didn't work. Just for some random chance I decided to try forward slashes and it started working. 

I don't know why that is the case but if anyone has an idea why that is I am very curious why that is.

**Reply by cooperdk** [reply]
DOUBLE backslash. It works. Both in webui-user.bat and in the interface.

**Reply by jfosterdesigns** [reply]
Forward slashes ( / ) work for me. Perhaps, double backslashes and forward slashes both work?

**Reply by cooperdk** [reply]
> Forward slashes ( / ) work for me. Perhaps, double backslashes and forward slashes both work?

Exactly, that is standard Python Directory specifications in Windows.

**Comment by Edu2014** [comment]
Is there any way to make multiple directories work? this way for example doesn't work  


![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/8563692/303ed972-e41d-4276-9001-ed2b9c57e992)

**Reply by cooperdk** [reply]
This has been discussed a million times.
Go to the default models folder.
Then create a symbolic link with the mklink command in that folder, one for each of your model directories wherever they reside. From inside the models folder, you would make a link to fx M:\models like this:

```mklink /D M-Drive M:\models```
You will then see a directory inside the models directory named M-Drive and clicking it takes you to the drive and directory you linked to. Do that for each drive and directory you store models.
Inside the web ui, the list of models will populate fully, sorted by the name of the links you created.

**Reply by Edu2014** [reply]
yes, I'm using this mklink /d method, but I have A LOT of folders.

**Reply by cooperdk** [reply]
Me too, but it's easy to manage.

**Comment by ostap667inbox** [comment]
As far as I know, the directory path should only be enclosed in quotes if it contains spaces. I have no spaces in the paths and they are written in webui-user.bat without quotes - and everything works.
Also, there is absolutely no condition to necessarily make the --ckpt-dir parameter first in the set COMMANDLINE_ARGS line. This line just lists the parameters, they are not ranked by importance in any way.

**Reply by cooperdk** [reply]
Once again, the way to do it has always been to make symbolic links in the model directory. This is just a solution that was implemented later.
The link method is much easier to manage.

**Reply by ostap667inbox** [reply]
I was only talking about quotation marks.
Everyone chooses a way that is convenient for them. Nothing to argue about. There is no "right" or "wrong" way to do it, as long as both work.

**Reply by cooperdk** [reply]
There was only one way until recently, and it worked perfectly. If you check all the discussions about it, linking has become the conventional method.

**Reply by ostap667inbox** [reply]
Do you write for the sake of the writing process itself? 
The ability to specify a directory for models and other files has been available in A1111 from the very beginning. Here is the relevant wiki page from Oct 18, 2022.
https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Command-Line-Arguments-and-Settings/cc7f49a020aa2061aac07220d91b370b4909276d
(There is no option for custom directory for LoRA because there was no LoRA yet in A1111 at that time)

**Reply by cooperdk** [reply]
It was only possible to specify ONE directory up until recently. For that reason, the recommended methods has always been linking. It was not even made public that it had become possible to specify multiple, as far as I can recall.
I wouldn't call the method easier, either.

BTW, the software had been available for months before October 22

**Comment by van1027** [comment]
Is it possible to change checkpoints and loras directories for linux? For some reason it doesn't find any models. I converted .bat files into .sh scripts, I tried to specify paths in the COMMANDLINE_ARGS and as separate variables, neither worked:

environment.sh:

```
#!/bin/bash

# Set the directory to the current script's directory with "system" subdirectory
DIR="$(dirname "$(realpath "$0")")/system"

# Update the PATH variable to include necessary directories
export PATH="$DIR/git/bin:$DIR/python:$DIR/python/Scripts:$PATH"

# Set additional Python-related environment variables
export PY_LIBS="$DIR/python/Scripts/Lib:$DIR/python/Scripts/Lib/site-packages"
export PY_PIP="$DIR/python/Scripts"
export SKIP_VENV=1
export PIP_INSTALLER_LOCATION="$DIR/python/get-pip.py"
export TRANSFORMERS_CACHE="$DIR/transformers-cache"

#export COMMANDLINE_ARGS="--ckpt-dir /mnt/backup/stable-diffusion-webui/models/Stable-diffusion --lora-dir /mnt/backup/stable-diffusion-webui/models/Lora"
export CKPT_DIR="/mnt/backup/stable-diffusion-webui/models/Stable-diffusion"
export LORA_DIR="/mnt/backup/stable-diffusion-webui/models/Lora"
export EMBEDDINGS_DIR="/mnt/backup/stable-diffusion-webui/embeddings"
export GFPGAN_DIR="/mnt/backup/stable-diffusion-webui/models/GFPGAN"
export VAE_PATH="/mnt/backup/stable-diffusion-webui/models/VAE"
export COMMANDLINE_ARGS="--xformers --cuda-malloc"

# Optional: print variables for verification
echo "Environment variables set:"
echo "PATH=$PATH"
echo "PY_LIBS=$PY_LIBS"
echo "PY_PIP=$PY_PIP"
echo "SKIP_VENV=$SKIP_VENV"
echo "PIP_INSTALLER_LOCATION=$PIP_INSTALLER_LOCATION"
echo "TRANSFORMERS_CACHE=$TRANSFORMERS_CACHE"
```
run.sh

```
#!/bin/bash

# Call the environment script
./environment.sh

# Change directory to "webui" relative to the script's location
cd ./webui

# Call the "webui-user.sh" script
./webui.sh
```
